In [71]:
from itertools import groupby
import pandas as pd

--- Day 1: Calorie Counting --- <br>
<br>
Santa's reindeer typically eat regular reindeer food, but they need a lot of magical energy to deliver presents on Christmas. For that, their favorite snack is a special type of star fruit that only grows deep in the jungle. The Elves have brought you on their annual expedition to the grove where the fruit grows.

To supply enough magical energy, the expedition needs to retrieve a minimum of fifty stars by December 25th. Although the Elves assure you that the grove has plenty of fruit, you decide to grab any fruit you see along the way, just in case.

Collect stars by solving puzzles. Two puzzles will be made available on each day in the Advent calendar; the second puzzle is unlocked when you complete the first. Each puzzle grants one star. Good luck!

---

The jungle must be too overgrown and difficult to navigate in vehicles or access from the air; the Elves' expedition traditionally goes on foot. As your boats approach land, the Elves begin taking inventory of their supplies. One important consideration is food - in particular, the number of Calories each Elf is carrying (your puzzle input).

The Elves take turns writing down the number of Calories contained by the various meals, snacks, rations, etc. that they've brought with them, one item per line. Each Elf separates their own inventory from the previous Elf's inventory (if any) by a blank line.

>Find the Elf carrying the most Calories. How many total Calories is that Elf carrying?

---

In [72]:
# Website requires log in for imput, but we access via other services, and log in for that service doesn't authenticate in r = requests.get(url, auth=('user', 'pass')) usagage.

# import requests
# try:
#     url = 'https://adventofcode.com/2022/day/1/input'
#     r = requests.get(url)
#     print("HTML:\n", r.text)
# except:
#     print("Invalid URL or some error occured while making the GET request to the specified URL")

In [73]:
with open('input.txt') as f:
    lines = f.readlines()

In [74]:
lines_test = lines

In [75]:
calories = []

for line in lines_test:
    cal = line.strip().strip('[').strip(']').split(',')
    calories.append(cal)

In [76]:
elf_cals = pd.DataFrame(calories)

In [77]:
elf_cals

0
0      9524
1     12618
2      6755
3      2121
4     12876
...     ...
2249   3910
2250   5347
2251   5630
2252   5721
2253   6630

[2254 rows x 1 columns]

In [78]:
clean_cals = list(elf_cals[0])

In [79]:
elf_try = [list(elf) for k, elf in groupby(clean_cals, key=bool) if k]

In [80]:
elf_df = pd.DataFrame(elf_try).T
# elf_df = pd.DataFrame(elf_try)

In [81]:
elf_df

0      1     2      3      4      5     6     7     8      9    ...  \
0    9524   4029  6226   5576   5523  32135  7380  9175  4807   1807  ...   
1   12618  11446  9901  23361  12113  15179  9715  4868  1180  13594  ...   
2    6755  11571  6735  16052   3128   None  5985  3374  6655  13712  ...   
3    2121   2683  7602   None   5978   None  9547  6053  4572   7385  ...   
4   12876   3454  4045   None   4683   None  9207  5304  3864   None  ...   
5   11330  12411  4643   None   3458   None  6882  4585  4674   None  ...   
6    None   None  5047   None   None   None  4407  1344  6759   None  ...   
7    None   None  2329   None   None   None  1688  5036  6861   None  ...   
8    None   None  None   None   None   None  None  None  4273   None  ...   
9    None   None  None   None   None   None  None  None  2387   None  ...   
10   None   None  None   None   None   None  None  None  1085   None  ...   
11   None   None  None   None   None   None  None  None  1598   None  ...   
12   None   None  None   None   None   None  None  None  None   None  ...   
13   None   None  None   None   None   None  None  None  None   None  ...   
14   None   None  None   None   None   None  None  None  None   None  ...   

     244   245    246    247   248   249   250    251   252   253  
0   5613  2815  45267  16209  3318  1854  7523   4172  2561  3059  
1   1938  7505   None  17734  1727  4926  3390  21931  3470  2276  
2   5481  3875   None   None  4971  3228  2318  22824  3018  1441  
3   3692  1236   None   None  5208  6199  8663   None  6387  3886  
4   5223  2028   None   None  8009  1312  7609   None  4774  2204  
5   1505  3022   None   None  8009  2417  1722   None  5611  3853  
6   3628  8262   None   None  6117  4413  5343   None  4904  3910  
7   1095  6240   None   None  7870  1801  8124   None  5883  5347  
8   4340  6295   None   None  4824  4145  1579   None  6144  5630  
9   2681  7060   None   None  7741  4604  2959   None  4977  5721  
10  5657  None   None   None  1278  4296  None   None  1863  6630  
11  1811  None   None   None  None  5636  None   None  1613  None  
12  1233  None   None   None  None  5169  None   None  3160  None  
13  None  None   None   None  None  6081  None   None  4236  None  
14  None  None   None   None  None  None  None   None  None  None  

[15 rows x 254 columns]

In [82]:
# elf_df.columns = ["elf1", "elf2","elf3", "elf4", "elf5", "elf6", "elf7", "elf8", "elf9", "elf10", "elf11", "elf12", "elf13", "elf14", "elf15"]

In [83]:
elf_df.isnull()

0      1      2      3      4      5      6      7      8      9    ...  \
0   False  False  False  False  False  False  False  False  False  False  ...   
1   False  False  False  False  False  False  False  False  False  False  ...   
2   False  False  False  False  False   True  False  False  False  False  ...   
3   False  False  False   True  False   True  False  False  False  False  ...   
4   False  False  False   True  False   True  False  False  False   True  ...   
5   False  False  False   True  False   True  False  False  False   True  ...   
6    True   True  False   True   True   True  False  False  False   True  ...   
7    True   True  False   True   True   True  False  False  False   True  ...   
8    True   True   True   True   True   True   True   True  False   True  ...   
9    True   True   True   True   True   True   True   True  False   True  ...   
10   True   True   True   True   True   True   True   True  False   True  ...   
11   True   True   True   True   True   True   True   True  False   True  ...   
12   True   True   True   True   True   True   True   True   True   True  ...   
13   True   True   True   True   True   True   True   True   True   True  ...   
14   True   True   True   True   True   True   True   True   True   True  ...   

      244    245    246    247    248    249    250    251    252    253  
0   False  False  False  False  False  False  False  False  False  False  
1   False  False   True  False  False  False  False  False  False  False  
2   False  False   True   True  False  False  False  False  False  False  
3   False  False   True   True  False  False  False   True  False  False  
4   False  False   True   True  False  False  False   True  False  False  
5   False  False   True   True  False  False  False   True  False  False  
6   False  False   True   True  False  False  False   True  False  False  
7   False  False   True   True  False  False  False   True  False  False  
8   False  False   True   True  False  False  False   True  False  False  
9   False  False   True   True  False  False  False   True  False  False  
10  False   True   True   True  False  False   True   True  False  False  
11  False   True   True   True   True  False   True   True  False   True  
12  False   True   True   True   True  False   True   True  False   True  
13   True   True   True   True   True  False   True   True  False   True  
14   True   True   True   True   True   True   True   True   True   True  

[15 rows x 254 columns]

In [84]:
elf_df = elf_df.fillna(0)

In [85]:
elf_df = elf_df.astype(int)

In [86]:
elf_sums = pd.Series(elf_df.sum())

In [87]:
elf_sums.max()

73211

First Answer is 73211 for my set of elves! That was probably a lot more complicated than needed, but I'm happy to have gotten to the answer. The hardest part was just getting the data in a format I could work with personally.

----

Now, PART 2:

--- Part Two --- <br>
By the time you calculate the answer to the Elves' question, they've already realized that the Elf carrying the most Calories of food might eventually run out of snacks.

To avoid this unacceptable situation, the Elves would instead like to know the total Calories carried by the top three Elves carrying the most Calories. That way, even if one of those Elves runs out of snacks, they still have two backups.

In the example above, the top three Elves are the fourth Elf (with 24000 Calories), then the third Elf (with 11000 Calories), then the fifth Elf (with 10000 Calories). The sum of the Calories carried by these three elves is 45000.

Find the top three Elves carrying the most Calories. How many Calories are those Elves carrying in total?

In [88]:
first_elf = elf_sums.max()
first_elf

73211

In [89]:
elf_sums[elf_sums == first_elf]

35    73211
dtype: int64

In [90]:
first_index = elf_sums[elf_sums == first_elf].index

In [91]:
first_index

Int64Index([35], dtype='int64')

In [92]:
elf_sums = elf_sums.drop(first_index)

In [93]:
second_elf = elf_sums.max()
second_elf

71575

In [94]:
second_index = elf_sums[elf_sums == second_elf].index

In [95]:
elf_sums = elf_sums.drop(second_index)

In [96]:
third_elf = elf_sums.max()
third_elf

69172

In [97]:
first_elf + second_elf + third_elf

213958